README.md
# spatialdemography
A spatially-explicit metacommunity model for R

You can install this package directly into R using the devtools package:
\# Install devtools from CRAN
install.packages("devtools")
\# Load devtools package
library(devtools)
\# Install this package from github
devtools::install_github(akeyel/spatialdemography)
\# Note: It appears that the vignettes will not be accessable through
\# browseVignettes("spatialdemography") unless you have the knitr package
\# (on CRAN) and pandoc (pandoc.org) installed.
\# However, the source is available in the vignettes folder.
 
See Description file for more details about the package.


NEWS
KNOWN BUGS:
None

KNOWN ISSUES:
No option to change rounding in calculation of UTC response trait diversity
Treatment of input paths for landscapes in the code requires improvement & increased clarity
Need to figure out how to not have the vignette code in two places (vignette folder & inst/docs folder)
Semi-colon as delimiter in code makes .csv files difficult to open in Excel on German computers as they use the ";" instead of the "," to separate fields.
Dispersal table warning message is not intuitive to someone who has never experienced the problem before. More description and more useful description is needed.

POTENTIAL FUTURE IMPROVEMENTS:
Provide option to allow dropping fractional individuals to prevent species 'resurrection'
Allow dispersal at stages other than 0
Allow more than 4 stages
Include juveniles in the competition with adults (or an option to do so)
Add quantile regression options for output
Add checks to trait calculation to ensure that vital rates are within the allowable range.
Drop landscape.identifiers and just use the entire environmental layer name in the code (would allow two environmental variables to start with the same name)

CHANGE LOG
# Changes in 1.0.1
Added option to enforce a pseudo-extinction threshold
Minor corrections to documentation

# Changes in 1.0.0
Updated NEWS to contain a list of potential future code improvements
Changed version number to 1.0.0 as part of a manuscript submission requirement

# Changes in 0.10.6
Removed extraneous code from vignettes
Added lifecycle graph to Vignette
Cleaned up comments from Sten in spatialdemography.r file

# Changes in 0.10.5
Fixed problem where landscape.dir would return an error if an entire landscape is input.
Added documentation to indicate that a pre-existing R object landscape can be used as an input
Matrix dependency for transpose function used on sparse matrices made explicit in code.
Changed code to automatically, recursively add required output paths.

#Changes in 0.10.4
Added the ability to process a threshold response to environmental variables (response function = 106)

#Changes in 0.10.3
Added stochastic lottery model as a competitiontype option
Began using Git for version control and collaborative development with Sten.
Added Require statments to beginning of SpatialDemography to check for required packages
Cleaned up formatting and fixed typos in Vignettes.

#Changes in 0.10.2
Removed Flags from code and dealt with flagged issues where appropriate
Flags were left that pertained to optimization
Updated timing file vignette to correspond to file run on cluster for the manuscript
Updated documentation for env.type for landscape setup
Added additional options to environmental change (all options used as distribution functions. But not yet available for conditional assignment.)
Removed (outdated) index of functions at beginning of sdhelper.r - the man files create a more up-to-date index
Patched assign.spp.v2 to work better with the special case of one unique species needs to be assigned to each cell.
Added aliases for sp.file documentation

#Changes in 0.10.1
Fixed a bug where the vignettes were not building properly.

#Changes in 0.10.0
Changed package name to spatialdemography to avoid name overlap with metacom package.
Updated test files to accomodate changed results file format
Added a vignette to illustrate the use of the package
Changed package to accept R objects as an inputs for initial.conditions.file, settings.file, env.file, and spfile. Also sp.instr.file and sp.resp.instr.file.
The directory for the environmental layers and the species locations file must now be entered as inputs and are no longer input through the settings file.
Landscape (Environmental layers) and species locations may also be input as R objects
The old locations file format is no longer supported by spatialdemography.
Model timing options and outputs changed
Fixed the bug "A landscape layer that changes may read in appropriately, but may not write to the same location, resulting in multiple files, one of which lacks header info (e.g., see Example 6 from the MS)"
   The landscape layer will write to a new file (still no header, sorry!) to avoid corrupting the original file.
Corrected minor bugs during testing

#Changes in 0.9.0
Changed format of results file
Added a warning message to let user know that dispersal tables are being generated and that this is a time consuming process.
Changed species initialization part of code - moved location in code, changed options and inputs, and added new options.
Made changes to make the package CRAN-compatible

#Changes in 0.8.5
Added a function for converting a raster file to an environmental layer input
Added a function to help generate the cell information for the Environmental Layers file

#Changes in 0.8.4
Added option to enter "no.lbl" for landscape generation to have an existing landscape with no additional label part.
Minor code cleanup
Functions from stottmatrix.r were moved to mchelper.r with permission from Iain Stott. These will be replaced by the functions in popdemo, once popdemo accepts sparseMatrix formats
stottmatrix.r is no longer required.

#Changes in 0.8.3
Alpha code for quantile regression & boosted regression trees for analyzing outputs (for Stochastic plants paper)
Fixed way model outputs were written to avoid code slowup bug introduced in version 0.8.0.
Fixed timing bug so that setup.matrices could be timed.
Added timing to compute.diagnostics - this step appears to be running extremely slowly for large landscapes.
Minor change to simulation2.r - removed some extraneous code where empty lists were created and then appended to - just wrote directly to create a new list.

#Changes in 0.8.2
Added na.omit to a step checking proportion of dispersing seeds - cells not producing seeds were causing an error in this calculation

#Changes in 0.8.1

inst/test/functiontests.r
# Purpose is to save code used to test code during development

#I'm sure there is a better way to do this!

# A.C. Keyel, begun 2015-03-04

#Species Loop - testing revisions

land.sp = sample(seq(1,16), 16)
patch.sp.num = 1
p = 16

#Test passed - each cell has a unique species and species check comes out correct.

#Check with 2 species per patch
land.sp = sample(seq(1,32), 32)
patch.sp.num = 2
p = 16


## Test new subsetting to just one environmental layer value type
#test of scl.subset
scale.cells.lst = "subset;L;1"
env.lbl = c("L","E")
landscape = list(c(1,2,2,1,1,2), c(0,1,2,3,4,5) ) 

scl.subset(scale.cells.lst, env.lbl, landscape)  
#Test passed - all 2's are now 0's

#Test with more values
scale.cells.lst = "subset;E;3"
scl.subset(scale.cells.lst, env.lbl, landscape)  
# Test passed - data were correctly recoded.

inst/test/test_spatialdemography.r
# CODE FOR TESTING THAT THE MODEL IS FUNCTIONING PROPERLY IN TESTABLE RESPECTS

#Main body of code borrowed from MC_model.r

#**# Think about converting the test code to vignettes?

#**# REVELATION: ../ refers to the parent directory with relative paths.  THIS IS COOL!  And probably utterly obvious to anyone who is actually a programmer...

#Tests should test all options, but not necessarily all interactions of options (because that would be a lot of testing)

#**# Figure out how to remove path dependency (will be easier for non-development mode code)
#**# Need to add clean up code that deletes the generated files
#**# Would file deletion be easier if the base code were all living somewhere else, and got copied into the appropriate folder?)
  #**# make that happen.

#TestRun1:
  # Test species generation
  # Test landscape generation
  # Test conditional landscape generation
  # Test initial location assignment (numeric)
  # Test with species invasion
  # Test with environmental change
  # Edge type = TORUS
  # 10 species at regional
  # 5 species in landscape
  # 1 species local
  # competition type = 0
  # multispecies.K = 1

#TestRun2
  # Test reading in species from file
  # Test read in landscape from file
  # Test initial location assignment (from file)
  # No species invasion
  # Test with environmental change
  #Edge type = TORUS
  # 2 species regional, landscape, and local
  # competition.type = 1
  # multispecies.K = 0

#TestRun3 #**# STILL NEEDS TO BE SET UP
  # Read species from file
  # Test with 1 species
  # test with 1 environmental layer
  #Test with landscape extent of 1
  #Edge type = ABSORBING
 
#TestRun4 #**# STILL NEEDS TO BE SET UP - USE A RUN FROM STOCHASTIC PLANTS
  # Test copula function
  # Include more variation in vital rates


#TestRun5 #**# STILL NEEDS TO BE SET UP - Add when doing theoretical traits paper
  # Generate species
  # Generate landscape
  # Test initial location assignment (biomass)
  # Edge type = ABSORBING
  # competition.type = 2
  # 32 species regional,  16 landscape and 4 local
  # RTD: 8 RTD total, 4 landscape, 2 local (4 x 4 landscape)
  # Have multiple scenarios, have each scenario test an environmental change type


#**# ALSO NEED TO SET UP SOME TESTS THAT WILL FAIL
  #Do so below, and wrap with a try/catch

# Fail RTD generation
# Fail conditional landscape variable
# other important fails?

#**# How to make this more general?? E.g., figure out where R can read example files, and use these as examples
setwd("C:/docs/beplants/Scripts/spatialdemography/inst/test/")
DispPath = "dt/"

dev = 0

if (dev == 0) library(spatialdemography)
if (dev == 1){
    #For development mode
    source("../../R/spatialdemography.r")
    source("../../R/simulation.r")
    source("../../R/sdhelper.r")
    #library(multirich)
    library(Matrix)
    }

#Source functions that ought to be in popdemo package, but don't work there.
#source("../../stottmatrix.r") #**# No longer necessary - these functions are in the package now

RunLog = "RunLog.csv"
if (file.exists(RunLog)){
    unlink(RunLog)
    }

to.run = c("TestRun1","TestRun2") #,"TestRun3","TestRun4","TestRun5" 
run.lbls = c("tr1","tr2","tr3","tr4","tr5")
for (r in 1:length(to.run)){
#r = 2

    #Set up run name and labels
    run.name = to.run[r]
    run.lbl = run.lbls[r]

    # Delete any previous test-run folder/files
    if (file.exists(run.name)){
        cat(sprintf("Old test run data for %s deleted\n", run.name), file = RunLog, append =T)
        unlink(run.name, recursive = T)
        }

    # Copy set up files into the appropriate location(s)
    # get listing of previous files
    base.dir = sprintf("testfiles/%s/", run.lbl)
    my.files = test.setup(base.dir, run.name)
      initial.conditions.file = my.files[[1]]
      settings.file = my.files[[2]]
      env.file = my.files[[3]]
      spfile = my.files[[4]]
      sp.instr.file = my.files[[5]]
      sp.resp.instr.file = my.files[[6]]
      locations.file = my.files[[7]]

    ## Run test & generate output ##
    # Create a path to store results & create results file
    file.ending = run.name #Could just put run name into the code, but this makes it faster when troubleshooting
    run.path = sprintf("%s/",run.name) #Set run path to run name
    opath = sprintf("%s/outputs/",run.name)
    dir.create(opath,showWarnings = F)    
    ResultsFile = sprintf("%sResults%s.csv",opath,file.ending)
    if (file.exists(ResultsFile)) file.remove(ResultsFile) #delete any existing results file

    landscape.dir = "default"
    if (r == 2){
        landscape.dir = "TestRun2/landscape/"
        }

    # set up scenarios #sc = 1 #**# Set up for more scenarios when testing multiple scenarios
    scn = 1 #scns[sc] #Scenario to run
    s.lbl = "" #s.lbls[sc]     #Label for the species file to be created

    #Set seed for random processes, to ensure consistent results
    set.seed(123456789) #Set seed, to allow consistent results from the random numbers generators.

    #Run SpatialDemography with testing mode (will load seeds to ensure repeatability
    out.res = SpatialDemography(scn,s.lbl,file.ending,DispPath,run.path,opath,ResultsFile,initial.conditions.file,settings.file,env.file,spfile, sp.instr.file,sp.resp.instr.file, landscape.dir = landscape.dir, locations.file = locations.file, testing = T)
    
    # Check generated output for accuracy & report any errors
    check.outputs(run.name,RunLog)
    
    # If no errors, delete test files
    cat(sprintf("%s run successfully with no errors. Deleting generated files.\n", run.name), file = RunLog, append = T)
    unlink(run.name, recursive = T)

    }                 



vignettes/Timing.R
## ------------------------------------------------------------------------
# Optionally set the working directory. The working directory should default to the vignettes directory. If it is changed, some of the inputs from the input files will need to be copied into the changed directory.
#setwd("C:/docs/beplants/Scripts/spatialdemography/vignettes") 

dev.mode = 1
if (dev.mode == 1){
    source("C:/docs/beplants/Scripts/spatialdemography/R/sdhelper.r")
    source("C:/docs/beplants/Scripts/spatialdemography/R/spatialdemography.r")
    source("C:/docs/beplants/Scripts/spatialdemography/R/simulation.r")
    library(Matrix)
    library(multirich)
}else{#Load the spatialdemography package
    library(spatialdemography)
    }

# Set up a RunName. Not actually an input, but we'll use it to build some of the required inputs
run.name = "spdem_timing"

# DispPath The path of the dispersal tables (if existing) or the path where dispersal tables should be created.
# Note that dispersal tables may take a long time to generate, so it is often desirable to share the same dispersal tables across multiple models.
DispPath = "dt/"

# run.path The main path for the specific overall model run.
run.path = sprintf("%s/",run.name) 

#The path for model outputs.
opath = sprintf("%soutputs/",run.path)

#Create the output path
dir.create(opath,showWarnings = F, recursive = T)

# A file to be created containing the model results.
# If there is an existing file, the results will be appended to the existing file.
ResultsFile = sprintf("%sResults_%s.csv",opath,run.name)

## Set up timing scenarios

# Set up the levels of variables for evaluation of model timing
species.vec = c(1,10,100) #1, 10, and 100 species
extent.vec = c(1,4,20) # 1 cell, 16 cells, 400 cells
change.vec = c(0,5,1) # No change, change every 5 timesteps, change every single timestep
timestep.vec = c(1,11,21) #One timestep, 11 timesteps, 21 timesteps (increment by 10)
vdb.vec = c(0,1) # Whether or not to generate many of the outputs in the model
matrix.diagnostics.vec = c(0,1) #Whether or not to generate many of the matrix diagnostic outputs

# Set default values for evaluation of other variables
spv = 10
exv = 4
chv = 5
tsv = 5
vdv = 0
mdv = 0

#Get the number of scenarios to be run
num.scn = length(species.vec) + length(extent.vec) + length(change.vec) + length(timestep.vec) + length(vdb.vec) + length(matrix.diagnostics.vec)

# Create an index of scenario numbers
in.seq = seq(1,num.scn)

# Set up vectors of inputs to be parallel with one another
sp.vec.pre = c()
sp.vec.val = species.vec #Create a vector of species values, with most scenarios run for 10 sp.
sp.vec.post = rep(spv, (num.scn - length(sp.vec.val)))
sp.vec = c(sp.vec.pre,sp.vec.val,sp.vec.post)

extent.vec.pre = rep(exv,length(sp.vec.pre) + length(sp.vec.val))
extent.vec.val = extent.vec
extent.vec.post = rep(exv, (num.scn - length(extent.vec.pre) - length(extent.vec.val)))
ex.vec = c(extent.vec.pre, extent.vec.val, extent.vec.post)

ch.vec.pre = rep(chv,num.scn - length(extent.vec.post))
ch.vec.val = change.vec
ch.vec.post = rep(chv, (num.scn - length(ch.vec.pre) - length(ch.vec.val)))
ch.vec = c(ch.vec.pre,ch.vec.val,ch.vec.post)

ts.vec.pre = rep(tsv, num.scn - length(ch.vec.post))
ts.vec.val = timestep.vec
ts.vec.post = rep(tsv, num.scn - length(ts.vec.pre)- length(ts.vec.val))
ts.vec = c(ts.vec.pre,ts.vec.val,ts.vec.post)

vd.vec.pre = rep(vdv, num.scn - length(ts.vec.post))
vd.vec.val = vdb.vec
vd.vec.post = rep(vdv, num.scn - length(vd.vec.pre) - length(vd.vec.val))
vd.vec = c(vd.vec.pre,vd.vec.val,vd.vec.post)

md.vec.pre = rep(mdv, num.scn - length(vd.vec.post))
md.vec.val = matrix.diagnostics.vec
md.vec.post = rep(mdv, num.scn - length(md.vec.pre) - length(md.vec.val))
md.vec = c(md.vec.pre,md.vec.val,md.vec.post)

#Check that vectors were correctly constructed. Note how only one vector has non-default values at once.
sp.vec
ex.vec
ch.vec
ts.vec
vd.vec
md.vec


## Set up initial.conditions and settings using Example1 as a template.
#Set the path where your inputs are stored
input.path = sprintf("spdem_ex1/inputs/")
ex1.ending = "_spdem_ex1"

# Read in the initial conditions file
initial.conditions.file = sprintf("%sInitial_conditions%s.csv",input.path,ex1.ending)
ic.ex1 = read.csv(initial.conditions.file)
#head(ic)

# Read in the settings file
settings.file = sprintf("%sSettings%s.csv",input.path,ex1.ending)
set.ex1 = read.csv(settings.file)
#head(set)

## Set up the other input files: env.file, species.instr.file, species.resp.file

#Read in environmental layers file
input.path = sprintf("spdem_timing/inputs/") #Change input path to correspond to the current folder.
env.file = sprintf("%sEnvironmental_layers_%s.csv",input.path,run.name)
ef = read.csv(env.file)
#ef #This will display the values of the environmental layers file
#?env.file #This will provide more details on the environmental layers file

#Read in species instructions file
sp.instr.file = sprintf("%sSpecies_instructions_spdem_timing.csv",input.path)
sif = read.csv(sp.instr.file)
#sp.instr.file  #This will display the values for the timing scenario 
#?sp.instr.file #This will provide more details on the format and codes used in the species instructions file

#Read in species response instructions file
sp.resp.instr.file = sprintf("%sSpecies_response_instr_spdem_timing.csv", input.path)
srif = read.csv(sp.resp.instr.file)
#srif #This will display the species response file values
#?sp.resp.instr.file #This will give more details on what the species response file values mean.

# Modify the base settings to match the timing scenario requirements
ic.ex1$num.adults = 1000 #Add num.adults field and initialize with 1000 of each species

set.ex1$GenerateNewSpecies = 1
set.ex1$GenerateNewLandscape = 1
set.ex1$loc.extraction = NULL
set.ex1$timestep.extraction = NULL
set.ex1$locations = NULL
set.ex1$ignore.rtd = 1
set.ex1$Write.Timefile = 1 #Only output the total model run time (2 would also reveal which sections of the model scale the worst) #**# Add to cluster version & set to 2 there!

ic = ic.ex1
set = set.ex1
#Expand the dataframes to contain the requisite number of scenarios (environmental layers does not need to be expanded)
for (i in 2:num.scn){
    ic = rbind(ic,ic.ex1)
    set = rbind(set,set.ex1)
    }

#Update the dataframes with the vectors
ic$Scenario = in.seq
ic$ModelName = sprintf("Timing%s",in.seq)
ic$Extent = ex.vec
ic$MaxTime = ts.vec
#set total species richness to equal landscape species richness to equal patch species richness.
#Not ecologically realisitc, but this model is being run for timing purposes, and not ecological realisim.
ic$Tot.sp.rich = ic$Land.sp.rich = ic$Loc.sp.rich = sp.vec  

set$Scenario = in.seq
set$GenerateVisualDebuggerData = vd.vec
set$RunMatrixDiagnostics = md.vec
set$lnd.lbl = sprintf("scn%s",in.seq)


#Code to (optionally) delete previous versions of the ResultsFile and TimeFile.
#**#
do.del = 1
if (do.del == 1){
  unlink(ResultsFile) #Delete any previous ResultsFile
  }

#Create a function to apply on the cluster
do.spdem = function(x, DispPath,run.path,opath,ResultsFile,ic,set,ef,sp.instr.file,sp.resp.instr.file,ch.vec){
    library(spatialdemography)
    scn = x 

    # s.lbl A label used in the construction of the Species file and new landscape files.
    s.lbl = sprintf("scn%s",x)

    # file.ending A text string that matches the end of the species file (either to be generated or that already exists).
    file.ending = sprintf("_%s_%s", run.name, s.lbl)
    
    #Update env.file to correspond to current change scenario
    ef$env.change.freq[1] = ch.vec[x]
    # Note: most scenarios will have more than one environmental layer, so it is not possible to structure the environmental
    # layers file in the same way as the initial.conditions file or the settings file. Hence why this data frame is updated here.

    # Create a file name & directory for the species file
    sp.path = sprintf("%sSpecies/",run.path)
    dir.create(sp.path,showWarnings = F, recursive = T)
    spfile = sprintf("%sspecies_file%s.csv", sp.path, file.ending)
    
    # set.seed function makes the stochastic results deterministic for repeatability purposes
    set.seed(567891234)

    #Run SpatialDemography. This is the main command to run the model, and uses the inputs defined above.
    SpatialDemography(scn,s.lbl,file.ending,DispPath,run.path,opath,ResultsFile,ic,set,ef,spfile,sp.instr.file,sp.resp.instr.file)    
    }

cpu = "PC"
#cpu = "cluster"

if (cpu == "cluster"){
    cores = 1
    workspace = "Timing_prerun.RData"
  
    library(parallel)
    workers      <- makeCluster(getOption("cl.cores", cores))     # Start worker processes
    save.image("Timing_prerun.RData")
  
    clusterEvalQ(workers, load("Timing_prerun.RData")) # Add workspace to each worker
    clusterApplyLB(cl=workers, x= in.seq, fun=do.spdem, DispPath,run.path,opath,ResultsFile,ic,set,ef,sp.instr.file,sp.resp.instr.file,ch.vec) # This is the main call for the analysis
  
    stopCluster(cl=workers)    # Once all analyses are done the workers are closed
}else{
    #Only run once if not running it on a cluster. This will prevent the vignette from running all scenarios automatically.
    #If not using a cluster, 1:1 can be replaced by in.seq. That will run all scenarios.
    for (i in 1:1){
        do.spdem(i, DispPath,run.path,opath,ResultsFile,ic,set,ef,sp.instr.file,sp.resp.instr.file,ch.vec)
        }
    }
  
#Timefile 
#This file records the timing of different steps of the model, and can be used for optimization purposes.
TimeFile = sprintf("%sTimefile.csv",opath)

timing = read.csv(TimeFile)
knitr::kable(timing)



vignettes/Example1.R
## ------------------------------------------------------------------------
# Optionally set the working directory. The working directory should default to the vignettes directory. If it is changed, some of the inputs from the input files will need to be copied into the changed directory.
#setwd("C:/docs/beplants/Scripts/spatialdemography/vignettes") 

#Load the spatialdemography package
library(spatialdemography)

# Set up a RunName. Not actually an input, but we'll use it to build some of the required inputs
run.name = "spdem_ex1"

## Set up required inputs

# scn = indicator variable to denote scenario and to extract appropriate row from initial.conditions.file and settings.file.
# If only one scenario is run, this should be 1, otherwise it should be updated inside a loop with one SpatialDemography command for each scenario.
scn = 1 

# s.lbl A label used in the construction of the Species file and new landscape files.
# Here we will leave it blank, but it may be relevant when running multiple scenarios.
s.lbl = ""

# file.ending A text string that matches the end of the species file (either to be generated or that already exists).
file.ending = sprintf("_%s", run.name)

# DispPath The path of the dispersal tables (if existing) or the path where dispersal tables should be created.
# Note that dispersal tables may take a long time to generate, so it is often desirable to share the same dispersal tables across multiple models.
DispPath = "dt/"

# run.path The main path for the specific overall model run.
run.path = sprintf("%s/",run.name) 

#The path for model outputs.
opath = sprintf("%soutputs/",run.path)

#Create the output path
dir.create(opath,showWarnings = F, recursive = T)

# A file to be created containing the model results.
# If there is an existing file, the results will be appended to the existing file.
ResultsFile = sprintf("%sResults%s.csv",opath,file.ending)

#Code to (optionally) delete previous versions of the ResultsFile and TimeFile.
#**#
do.del = 1
if (do.del == 1){
  unlink(ResultsFile) #Delete any previous ResultsFile
  }

## ------------------------------------------------------------------------
#Set the path where your inputs are stored
input.path = sprintf("%sinputs/",run.path)

# initial.conditions.file A file to specify the initial model conditions.
initial.conditions.file = sprintf("%sInitial_conditions%s.csv",input.path,file.ending)

# Here we will read in the file, to look at its contents
ic = read.csv(initial.conditions.file)

## ------------------------------------------------------------------------
knitr::kable(head(ic[1:5]))

## ------------------------------------------------------------------------
knitr::kable(head(ic[6:9]))

## ------------------------------------------------------------------------
knitr::kable(head(ic[10:13]))

## ------------------------------------------------------------------------
knitr::kable(head(ic[14:20]))

## ------------------------------------------------------------------------
# settings.file A file to specify overall model settings.
settings.file = sprintf("%sSettings%s.csv",input.path,file.ending)

set = read.csv(settings.file)
knitr::kable(head(set[1:5]))

## ------------------------------------------------------------------------
knitr::kable(head(set[6:7]))

## ------------------------------------------------------------------------
knitr::kable(head(set[8:12]))

## ------------------------------------------------------------------------
# env.file A file to describe environmental layers and how they change.
env.file = sprintf("%sEnvironmental_layers%s.csv",input.path,file.ending)

ef = read.csv(env.file)
knitr::kable(head(ef[1:2]))

## ------------------------------------------------------------------------
knitr::kable(head(ef[3:8]))

## ------------------------------------------------------------------------
knitr::kable(head(ef[9:11]))

## ------------------------------------------------------------------------
lpath = "spdem_ex1/landscape/"
lc.file = sprintf("%slandcover_scn1.csv",lpath)
h.file = sprintf("%sherbicide_scn1.csv",lpath)

lc = read.csv(lc.file)
knitr::kable(lc)
#Above is the basic file structure

lc2 = matrix(lc[1,2:length(lc)], nrow = 2, byrow = TRUE) #starts at 2 to avoid the change.step
lc2
#Above is the layer in landscape format for the first change step (landcover in this example only has one change step)

#We will now examine the herbicide layer:
#Please note that the change.step in the landscape files is equal to the timestep + 1, as the change takes effect in the NEXT model run
#Forest cells (2) are not treated with herbicides.

#Herbicides
#The herbicides layer has many change steps specified because it does not change predictably at a regular interval (i.e. herbicides are applied twice, and then the layer reverts back to no application.)
herb = read.csv(h.file)
knitr::kable(herb)

## ------------------------------------------------------------------------
# spfile A file specifying species traits.
# This file should be in the folder Species (it is not in the inputs folder because sometimes a species file is an output of SpatialDemography)
spfile = sprintf("%sSpecies/Species%s.csv",run.path, file.ending)

# sp.instr.file A file to describe how generation of species' base vital rates is to occur
# Optional file only needed if SpatialDemography is generating species so we'll set it to NA (it can also be absent entirely)
sp.instr.file = NA      

# sp.resp.instr.file A file to describe how generation of species' response traits is to occur
# Optional file only needed if SpatialDemography is generating species, so we'll set it to NA (it can also be absent entirely)
sp.resp.instr.rile = NA #Optional file only needed if SpatialDemography is generating species

#Examine the species file
spf = read.csv(spfile)
knitr::kable(head(spf))

## ------------------------------------------------------------------------
locations.file = sprintf("%slocations%s.csv", input.path, file.ending)
locs = read.csv(locations.file)
knitr::kable(locs)


## ------------------------------------------------------------------------
#**# Only run when troubleshooting code. Delete from final vignette
dev.mode = 1
if (dev.mode == 1){
  source("C:/docs/beplants/Scripts/spatialdemography/R/sdhelper.r")
  source("C:/docs/beplants/Scripts/spatialdemography/R/spatialdemography.r")
  source("C:/docs/beplants/Scripts/spatialdemography/R/simulation.r")
  library(Matrix)
  library(multirich)
  }

restart = 0
if (restart == 1){
  myworkspace = "TS.RData"
  load(myworkspace)
  }


## ------------------------------------------------------------------------
# set.seed function makes the stochastic results deterministic for repeatability purposes
# Not actually relevant for this example, as there are no stochastic processes included.
set.seed(567891234)

#Run SpatialDemography. This is the main command to run the model, and uses the inputs defined above.
SpatialDemography(scn,s.lbl,file.ending,DispPath,run.path,opath,ResultsFile,initial.conditions.file,settings.file,env.file,spfile,sp.instr.file,sp.resp.instr.file,lpath,locations.file)


## ------------------------------------------------------------------------
#Results file
out.results = read.csv(ResultsFile)
knitr::kable(out.results)

## ------------------------------------------------------------------------
#Next, we will consider the scenario-specific files.
epath = "spdem_ex1/Example1/"

#SpeciesData.csv will contain species abundances of each lifestage (and all lifestages together) for every cell in the landscape.
sp.dat = read.csv(sprintf("%sSpeciesData.csv",epath))
knitr::kable(head(sp.dat))

#These data are summarized in the following files (in the Example1 folder):
#Cells_occupied.csv indicates how many cells are occupied by each species at each timestep
#SpeciesStats_Adults_v2.csv indicates how many adults of each species there is in the landscape for each timestep, and also the log of lambda (growth rate).
#SpeciesStats_All_v2.csv is the same as SpeciesStats_Adults_v2.csv, except for all life stages pooled.
#SpeciesStats_Juvs.csv indicates how many juveniles there are of each species in the landscape, but does not include loglambda values
#SpeciesStats_Seeds.csv as with juveniles, but for seeds

#One other file in this folder bears note: change_count_lookup.csv provides an index for which change steps correspond to which timesteps (because a change in environmental conditions may be less frequent than a change in timesteps).


#Finally, we will consider the diagnostic files.
dpath = sprintf("%sDiagnostics/", epath)

#There are four files in this folder:
# AMatrices.csv contains the entire matrix A for each species for each timestep. This can be useful for checking that vital rates are as intended and that the matrix was properly constructed.

#TransitionMatrices.csv gives the transition matrices within a cell for each cell in the landscape, and can be useful for examining local dynamics exclusive of immigration and emigration.
# AMatricesSummaries.csv summarizes the matrix content for each cell
mat.sum = read.csv(sprintf("%sAMatricesSummaries.csv",dpath))
knitr::kable(head(mat.sum))

#MatrixDiagnostics.csv summarizes the matrices across all cells for a species for each change step.
mat.dia = read.csv(sprintf("%sMatrixDiagnostics.csv",dpath))
knitr::kable(mat.dia)



## ------------------------------------------------------------------------
#We'll start by looking at the SpeciesData.csv a bit more closely.
#Specifically, we'll look at the adult numbers for each species (rounded for ease of interpretation)
sp.dat.ad = sp.dat[sp.dat$LifeStage == "Adults", ]
sp.dat.ad$LifeStage = NULL
#sp.dat.ad = sp.dat.ad[order(sp.dat.ad$Species), ]
sp.dat.ad1 = sp.dat.ad[sp.dat.ad$Species == 1, ]
sp.dat.ad2 = sp.dat.ad[sp.dat.ad$Species == 2, ]
sp.dat.ad3 = sp.dat.ad[sp.dat.ad$Species == 3, ]

knitr::kable(round(sp.dat.ad1,0)) #Examine Species 1 Adult data


## ------------------------------------------------------------------------
knitr::kable(round(sp.dat.ad2,0)) #Examine Species 2 Adult data

## ------------------------------------------------------------------------
knitr::kable(round(sp.dat.ad3,0)) #Examine Species 3 Adult data

## ----, fig.show = 'hold'-------------------------------------------------
#Graphically examine the overall patterns by species
# Read in SpeciesStats_Adults_v2.csv
ssa = "spdem_ex1/Example1/SpeciesStats_Adults_v2.csv"
my.df = read.csv(ssa)

x.limit = 15
y.limit = 20000

#Plot Species 1
plot(my.df$RunTime, my.df$Sp1, xlab = "Time Step", ylab = "Number of Adults", xlim = c(0,x.limit), ylim = c(0, y.limit), pch = 15, col = 4)

par(new = T) #Tell R to add next plot to existing plot

#Plot Species 2
plot(my.df$RunTime, my.df$Sp2, xlim = c(0,x.limit), ylim = c(0, y.limit), pch = 16, col = 2, xlab = "", ylab = "", xaxt = 'n', yaxt = 'n')

par(new = T) #Tell R to add next plot to existing plot

#Plot Species 3
plot(my.df$RunTime, my.df$Sp3, xlim = c(0,x.limit), ylim = c(0, y.limit), pch = 17, col = 3, xlab = "", ylab = "", xaxt = 'n', yaxt = 'n')

## Examine Species 3's performance by cell
x.limit = 15
y.limit = log(7000,10)

#Plot Cell1
plot(sp.dat.ad3$TimeStep,log(sp.dat.ad3$Cell1,10), xlim = c(0,x.limit), ylim = c(0,y.limit), pch = 1, col = "orange", xlab = "Time Step", ylab = "Log10 Number of Adults" )

#Plot Cells 2 - 4
for (i in 2:4){
    pch.vec = c(NA,6,20,0)
    par(new = T) #Tell R to add next plot to existing plot
    this.cell = sprintf("Cell%s", i)
    plot(sp.dat.ad3$TimeStep, log(sp.dat.ad3[[this.cell]],10), xlim = c(0,x.limit), ylim = c(0, y.limit), pch = pch.vec[i], col = i, xlab = "", ylab = "", xaxt = 'n', yaxt = 'n')
    }



